### DISCUSSION TITLE
Setting field metadata with BaseModel and Field

### DISCUSSION THREAD

**Discussion by eudoxos** [pydantic/pydantic#2419] [Question] [answered]
Hi, I am converting code using vanilla `dataclasses` decorator to pydantic. That code uses metadata to set some special attributes, such as that the field should not be serialized when the instance is transferred over the wire, like this:

```python
someField: typing.Any=dataclasses.field(default=None,metadata=dict('remote_serialize':False))
```
This works okay when using `pydantic.dataclasses.dataclass` instead of vanilla (as transition), but when I go a step further to use `pydantic.BaseModel` instead, does `Field` provide some comparable functionality? [Field customization](https://pydantic-docs.helpmanual.io/usage/schema/#field-customisation) does seem to mention any way of extending the information it carries.

Thanks for hints!

**Comment by eudoxos** [comment]
Okay, I think I got it; one needs to get `**` args of the `Field` from the schema, not from the Field itself:
```python
import pydantic
class A(pydantic.BaseModel):
   a: int=pydantic.Field(myattr=True)

# it is not here:
A.__fields__
# {'a': ModelField(name='a', type=int, required=True)}

# but it is in the schema
A.schema()['properties']['a']
# {'title': 'A', 'myattr': True, 'type': 'integer'}

```

**Comment by PrettyWood** [comment, marked as answer]
Hi @eudoxos 
IMO it's better to use `field_info` attribute of a field like this
```py
from pydantic import BaseModel, Field


class A(BaseModel):
   a: int = Field(myattr=True)

assert A.__fields__['a'].field_info.extra == {'myattr': True}
```

**Reply by eudoxos** [reply]
Thanks, this much better.

**Reply by ericfrederich** [reply]
FYI, using Pydantic v2 you get this warning.

> PydanticDeprecatedSince20: Using extra keyword arguments on `Field` is deprecated and will be removed. Use `json_schema_extra` instead. (Extra keys: 'pipeline_variable'). Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.6/migration/

**Comment by g0di** [comment]
Sorry for the revival but I ended up recently in an exact same situation but I think the provided answer is not   acceptable anymore since using  `**extra` is clearly marked as deprecated with v2 and is now intended for json schema, not python metadata.

At the moment I found two solutions to add metadata:

### Using `Annotated`

With the introduction of `Annotated` type, the most straightforward way to add metadata to a `Field` is by simply adding those metadata to the `Annotated` type which is completly designed for this case
```python
from typing import Annotated

from pydantic import BaseModel, Field


class Foo(BaseModel):
    bar: Annotated[str, {"meta": "data"}] = Field(...)


print(Foo.model_fields["bar"].metadata)
# [{'meta': 'data'}]
```
I'm not sure this is actually documented and it could be interesting to have it. Note that I personally find this solution a bit tedious when you start having a lot of models/fields on which you need to apply metadata. It can makes things a bit harder to read.

## Wrapping the `Field` function 
Another solution I've implemented for a personal library on top of pydantic is to provide my own `Field` function which wrap the original one and pass metadata to it.

```python
from pydantic import BaseModel, Field


def MyField(default, *, meta: str | None = None, **kwargs):
    field_info = Field(default, **kwargs)
    field_info.metadata.append({"meta": meta})
    return field_info


class Foo(BaseModel):
    bar: str = MyField(..., meta="data")


print(Foo.model_fields["bar"].metadata)
# [{'meta': 'data'}]
```
This way makes it easier to apply metadata without having to wrap all fields with `Annotated`. However, it comes with serious downsides:
- You have to copy/paste all pydantic `Field` arguments with their correct types plus the whole doc if you want to still have all typings, auto-completions, etc ... which can makes it really fragile on pydantic updates
- You have to create your own metaclass with proper `dataclass_transform` decorator in order to let your IDE knows that `MyField` is actually a field descriptor or your IDE will consider it as the default value of the field: 

![image](https://github.com/pydantic/pydantic/assets/2545380/1f595517-9ae0-44f4-b8a1-a9c4a985d152)


I think it could be interesting however to maybe allow passing metadata directly through the `Field` function to avoid ending with both `Annotated` and `Field` on the same fields.

**Reply by ericfrederich** [reply]
Thanks for this writeup.  I have a need to add some metadata to each of my fields.  This metadata has absolutely nothing to do with any JSON Schema, so it feels completely wrong to use `json_schema_extra`.

If your 2nd solution has typing implications then I want nothing to do with it ;-)  Half the reason I'm using Pydantic instead of JSON and dicts is to have better IDE support.

Your 1st solution is interesting but I'm wondering if its an abuse of type annotations.  Similar to how I don't want to just use `json_schema_extra` because it has nothing to do with JSON Schema... I wouldn't want to use Annotated if the information I'm putting in there isn't really anything related to types.

Also I have existing stuff in Annotated already (`Annotated[IsDigits, MinLen(4), MaxLen(4)]`, `Annotated[Optional[str], MaxLen(2000)]`, etc).  I'd have to iterate through the list returned by `model_instance.model_fields["field_name"].extra`.

Is my understanding of Annotated wrong?  Is this what it was meant for?

Specifically we have a model where we take in camel case variables from a yml/json file.  In the Python code we use Pythonic variable names, but we also need to export it to a file which ends up being used to create environment variables.  So for each field we need 3 names (i.e. `foo_bar`, `FooBar`, and `FOO_BAR`)

This is what we have but it again feels wrong to be abusing `json_schema_extra`

```python
class MyModel(BaseModel):
    foo_bar: Annotated[Optional[str], MaxLen(2000)] = Field(
        alias="FooBar", json_schema_extra={"env_var": "FOO_BAR"}
    )
```

**Reply by g0di** [reply]
Well the second solution requires a bit more work but you still end up with IDE support. However it is more brittle IMO because you're basically wrapping the Pydantic original `Field` function.

I think another solution would be to have a `metadata` keyword argument in the `Field` function for attaching custom metadata. However, using the `Annotated` stuff is mostly doing that. It actually bring annotations into the Pydantic `FieldInfo` `metadata` attribute so I guess this is the expected mechanism for attaching any kind of metadata to a Pydantic field.

```python
from typing import Annotated

from pydantic import BaseModel


class Foo(BaseModel):
    bar: Annotated[str, {"some": "metadata"}]


assert Foo.model_fields["bar"].metadata == [{"some": "metadata"}]
```


By the way Pydantic already provide `Annotated` support for customizing fields in various ways (json schema generation, validation, serialization, ...). Other libraries like FastAPI also provide ability to attach library metadata using `Annotated` (for dependency injection or documenting a route parameter as a query parameter) so I think its mostly ok to attach any kind of metadata using `Annotated` even if those are not closely related to the type being annotated. IMO, this is really the role of `Annotated`. 

The only thing I dislike regarding using `Annotated` is that:
-  it can makes thing lot more verbose
- Not everyone is comfortable using `Annotated` yet

For example, consider this very small example for, lets say, building an ORM on top of pydantic where I want to exclude one of my model attribute from the generated JSON schema (for reasons):

```python
from dataclasses import dataclass
from typing import Annotated

from pydantic import BaseModel, Field
from pydantic.json_schema import SkipJsonSchema


@dataclass
class Column:
    name: str
    type: str


class Foo(BaseModel):
    name: Annotated[str, Column("NAME", "STRING")]
    bar: SkipJsonSchema[
        Annotated[list[str], Field(default_factory=list), Column("BAR", "LIST[STRING]")]
    ]
```
This is working as expected but when you'll start having lot of modes with lot of attributes, custom validations/serializations and so on, things can get quite verbose.

**Reply by ericfrederich** [reply]
Wow, that SkipJsonSchema makes it real ugly.
Would look nicer if it were inside of the Annotated[] instead of encapsulating it.  Why wouldn't Pydantic support marking it to skip JSONSchema _inside_ of `Annotated`?

```python
class Foo(BaseModel):
    name: Annotated[str, Column("NAME", "STRING")]
    bar: Annotated[list[str], Field(default_factory=list), Column("BAR", "LIST[STRING]"), SkipJsonSchema]
```

**Reply by g0di** [reply]
Well actually I _think_ you can do this:
```python
bar: Annotated[list[str], Field(default_factory=list), Column("BAR", "LIST[STRING]"), SkipJsonSchema()]
```

**Comment by gislerro** [comment]
I think I've found a decent way to inject custom metadata into a model in a typesafe way:

![image](https://github.com/user-attachments/assets/0a879aa3-7204-4101-b822-f4cef5e7904d)

You can wrap pydantic's `Field` function to store your custom metadata and manually adjust the type signature:

```python
@dataclass
class MyMetadata:
    foo: str | None = None
    bar: bool = False


# Adapted from: https://stackoverflow.com/questions/71968447/python-typing-copy-kwargs-from-one-function-to-another
# also refer to discussion here: https://discuss.python.org/t/taking-the-argument-signature-from-a-different-function/42618/20
#
# FIXME: ...in the future - It's currently impossible to slot in 'MyMetadata' as a named kwarg:
# https://peps.python.org/pep-0612/#concatenating-keyword-parameters
#
P = ParamSpec("P")  # param spec of wrapper
T = TypeVar("T")  # return type of wrapped function


def wrap(
    _: Callable[P, Any]
) -> Callable[[Callable[..., T]], Callable[Concatenate[Optional[MyMetadata], P], T]]:
    """Wrap a `Converter` `__init__` in a type-safe way."""

    def impl(
        fun: Callable[..., T]
    ) -> Callable[Concatenate[Optional[MyMetadata], P], T]:
        return cast(Callable[Concatenate[Optional[MyMetadata], P], T], fun)

    return impl


@wrap(Field)
def MyField(mymetadata: MyMetadata | None = MyMetadata(), *args, **kwargs):
    """My custom docstring"""

    field = Field(*args, **kwargs)
    # in the original FieldInfo metadata is an array so keep it consistent
    field.metadata.append(mymetadata)

    # ideally inside your model you'd write a function to extract the metadata
    print(field.metadata[0])
    return field
    
if TYPE_CHECKING:

    class MyFieldInfo(FieldInfo):
        metadata: list[Any | MyMetadata]


class MyModel(BaseModel):

    if TYPE_CHECKING:
        # valid since MyFieldInfo is a subtype of FieldInfo
        model_fields: ClassVar[Dict[str, MyFieldInfo]]  # type: ignore

    @classmethod
    def __extract_mymetadata(cls, field_info: MyFieldInfo) -> Optional[MyMetadata]:
        for metadata in field_info.metadata:
            if isinstance(metadata, MyMetadata):
                return metadata
        return None

    @classmethod
    def get_mymetadata(cls) -> Dict[str, MyMetadata]:
        field_mymetadata = {}
        for field_name, field_info in cls.model_fields.items():
            mymetadata = cls.__extract_mymetadata(field_info)
            if mymetadata:
                field_mymetadata[field_name] = mymetadata
        return field_mymetadata

    @classmethod
    def __pydantic_init_subclass__(cls, **kwargs: Any) -> None:
        super().__pydantic_init_subclass__(**kwargs)
        mymetadata = cls.get_mymetadata()
        print(f"__pydantic_init_subclass__ MyMetadata: {mymetadata}")
        # do something useful


class Foo(MyModel):
    withmetadata: str = MyField(
        # my metadata is hidden to the json schema
        MyMetadata(foo="custom", bar=False),
        ...,
        # since extra is deprecated, this will be put into json_schema_extra
        foo={"implicitly into json schema"},
    )
    partialmetadata: str = MyField(MyMetadata(bar=True), ...)
    # None must be passed explicitly - since kwargs 'reshuffling' is not yet supported
    nometadata: str = MyField(None, default="my default", description="my description")
    
foo = Foo(withmetadata="foo", partialmetadata="bar")
print(foo)
pprint.pprint(foo.model_json_schema(mode="serialization"), indent=4)
```

which prints:

```bash
__pydantic_init_subclass__ MyMetadata: {'withmetadata': MyMetadata(foo='custom', bar=False), 'partialmetadata': MyMetadata(foo=None, bar=True)}


withmetadata='foo' partialmetadata='bar' nometadata='my default'
{   'properties': {   'nometadata': {   'default': 'my default',
                                        'description': 'my description',
                                        'title': 'Nometadata',
                                        'type': 'string'},
                      'partialmetadata': {   'title': 'Partialmetadata',
                                             'type': 'string'},
                      'withmetadata': {   'foo': [   'implicitly into json '
                                                     'schema'],
                                          'title': 'Withmetadata',
                                          'type': 'string'}},
    'required': ['withmetadata', 'partialmetadata'],
    'title': 'Foo',
    'type': 'object'}
```

This could be made more ergonomic if there's a way to shuffle kwargs: https://peps.python.org/pep-0612/#concatenating-keyword-parameters so MyMetadata could be moved to the kwargs
